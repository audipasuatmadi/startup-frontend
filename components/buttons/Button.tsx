import React from 'react';
import BaseButton from './BaseButton';

type themes = 'brand' | 'add later';

export interface ButtonProps {
  disabled?: boolean;
  href?: string;
  children?: any;

  className?: string;
  type?: 'button' | 'submit' | 'reset';

  onClick?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;
  onMouseDown?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;

  rounded?: boolean;
  outlined?: boolean;
  theme?: themes;
}

/**
 * 
 * @version 0.1.0
 * @description basically wraps BaseButton, adding `outlined` and `rounded` features.
 * @param rounded boolean
 * @param outlined boolean
 * @author Putu Audi Pasuatmadi 
 */
const Button: React.FC<ButtonProps> = ({
  outlined,
  rounded,
  theme,
  disabled,
  className,
  ...otherProps
}: ButtonProps) => {
  const finalStyles = `button${rounded ? ' rounded-full' : ''}${
    disabled
      ? ` button-disabled bg-${theme}-dark text-white`
      : `${
          outlined
            ? ` bg-none border-2 border-${theme} text-${theme} hover:text-white`
            : ` bg-${theme} text-white`
        } hover:bg-${theme}-dark active:bg-${theme}-light focus:ring-4 ring-${theme} ring-opacity-50`
  } ${className}`;

  return (
    <BaseButton
      className={`${finalStyles}`}
      disabled={disabled}
      {...otherProps}
    />
  );
};

Button.defaultProps = {
  theme: 'brand'
}

export default Button;
