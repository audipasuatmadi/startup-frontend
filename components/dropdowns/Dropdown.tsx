import React, { useEffect, useRef, useState } from 'react';

interface TogglerRenderProps {
  onClick: (e: React.MouseEvent) => void;
}

export interface DropdownProps {
  togglerRender: (props: TogglerRenderProps) => JSX.Element;

  containerAlign?: 'left' | 'center' | 'right';
  children?: JSX.Element | JSX.Element[];
}

export function useOnClickOutsideComponent(ref: React.MutableRefObject<any>, callbackFunction: () => void) {
  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (ref.current && !ref.current.contains(event.target)) {
        callbackFunction();
      }
    }

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [ref, callbackFunction]);
}

const getDropdownContainerStyling = (containerAlign: 'left' | 'center' | 'right') => {
  return containerAlign === 'center'
    ? 'w-full'
    : containerAlign === 'left'
    ? 'w-64 left-0'
    : 'w-64 right-0';
};


const Dropdown: React.FC<DropdownProps> = ({
  togglerRender,
  containerAlign,
  children,
}: DropdownProps) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleOpenClose = () => setIsOpen(!isOpen);
  const wrapperRef: React.MutableRefObject<any> = useRef(null);
  useOnClickOutsideComponent(wrapperRef, () => setIsOpen(false));

  return (
    <div className='relative inline-block' ref={wrapperRef}>
      {togglerRender({ onClick: handleOpenClose })}

      <ul
        className={`${getDropdownContainerStyling(
          containerAlign!
        )} bg-gray-200 rounded absolute shadow-md${
          isOpen ? '' : ' invisible'
        }`}
      >
        {children}
      </ul>
    </div>
  );
  
};

Dropdown.defaultProps = {
  containerAlign: 'center',
};

export default Dropdown;
