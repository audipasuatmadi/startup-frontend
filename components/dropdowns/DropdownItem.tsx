import React from 'react';
import TextButton from '../buttons/TextButton';

export interface DropdownItemProps {
  name?: string;
  type?: 'button' | 'submit' | 'reset';
  className?: string;
  children?: string;
  disabled?: boolean;

  onClick?: (
    e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>
  ) => void;
  href?: string;

  divider?: boolean
}

const DropdownItem = ({className, divider, ...otherProps}: DropdownItemProps) => {
  return (
    <li
      className={`list-none${
        divider? ' border-t-2 border-gray-300' : ''
      }`}>
      <TextButton
        className='block w-full'
        {...otherProps}
      />
    </li>
  );
};

export default DropdownItem;
