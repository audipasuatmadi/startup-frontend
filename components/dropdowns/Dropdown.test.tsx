import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import Button from '../buttons/Button';
import Dropdown from './Dropdown';
import DropdownItem from './DropdownItem';

describe('dropdown component render tests', () => {
  it('should render correctly without childrens, shallowy', () => {
    const wrapper: ShallowWrapper = shallow(
      <Dropdown
        togglerRender={(props) => <Button {...props}>Test Button</Button>}
      />
    );

    expect(wrapper.find('Button').first().exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with childrens, shallowy', () => {
    const component: JSX.Element = (
      <Dropdown
        togglerRender={(props) => <Button {...props}>Test Button</Button>}>
        <DropdownItem>Item 1</DropdownItem>
        <DropdownItem>Item 2</DropdownItem>
        <DropdownItem>Item 3</DropdownItem>
      </Dropdown>
    );

    const wrapper: ShallowWrapper = shallow(component);

    expect(wrapper.find('Button').first().exists()).toBe(true);
    expect(wrapper.find('DropdownItem').length).toEqual(3);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with childrens disabled, shallowy', () => {
    const component: JSX.Element = (
      <Dropdown
        togglerRender={(props) => <Button {...props}>Test Button</Button>}>
        <DropdownItem disabled onClick={() => {}}>
          Item 1
        </DropdownItem>
        <DropdownItem disabled>Item 2</DropdownItem>
        <DropdownItem>Item 3</DropdownItem>
      </Dropdown>
    );

    const wrapper: ShallowWrapper = shallow(component);

    expect(wrapper.find('Button').first().exists()).toBe(true);
    expect(wrapper.find('DropdownItem').length).toEqual(3);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with childrens, opened', () => {
    const component: JSX.Element = (
      <Dropdown
        togglerRender={(props) => <Button {...props}>Test Button</Button>}>
        <DropdownItem>Item 1</DropdownItem>
        <DropdownItem>Item 2</DropdownItem>
        <DropdownItem>Item 3</DropdownItem>
      </Dropdown>
    );

    const wrapper: ShallowWrapper = shallow(component);

    expect(wrapper.find('Button').first().exists()).toBe(true);
    wrapper.find('Button').simulate('click');

    expect(wrapper.find('DropdownItem').length).toEqual(3);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('dropdown component functionality test', () => {
  let testValue = false;

  const setTestValue = (
    e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>
  ) => {
    testValue = true;
  };

  beforeEach(() => {
    testValue = false;
  });

  it('should have DropdownItem and able to be clicked', () => {
    const component: JSX.Element = (
      <Dropdown
        togglerRender={(props) => <Button {...props}>Test Button</Button>}>
        <DropdownItem onClick={setTestValue}>Item 1</DropdownItem>
        <DropdownItem>Item 2</DropdownItem>
        <DropdownItem>Item 3</DropdownItem>
      </Dropdown>
    );

    const wrapper: ShallowWrapper = shallow(component);

    wrapper.find('DropdownItem').first().simulate('click');
    expect(testValue).toBe(true);
  });
});
