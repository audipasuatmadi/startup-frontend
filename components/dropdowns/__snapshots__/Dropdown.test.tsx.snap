// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown component render tests should render correctly with childrens disabled, shallowy 1`] = `
<div
  className="relative inline-block"
>
  <Button
    onClick={[Function]}
    theme="brand"
  >
    Test Button
  </Button>
  <ul
    className="w-full bg-gray-200 rounded absolute shadow-md invisible"
  >
    <DropdownItem
      disabled={true}
      onClick={[Function]}
    >
      Item 1
    </DropdownItem>
    <DropdownItem
      disabled={true}
    >
      Item 2
    </DropdownItem>
    <DropdownItem>
      Item 3
    </DropdownItem>
  </ul>
</div>
`;

exports[`dropdown component render tests should render correctly with childrens, opened 1`] = `
<div
  className="relative inline-block"
>
  <Button
    onClick={[Function]}
    theme="brand"
  >
    Test Button
  </Button>
  <ul
    className="w-full bg-gray-200 rounded absolute shadow-md"
  >
    <DropdownItem>
      Item 1
    </DropdownItem>
    <DropdownItem>
      Item 2
    </DropdownItem>
    <DropdownItem>
      Item 3
    </DropdownItem>
  </ul>
</div>
`;

exports[`dropdown component render tests should render correctly with childrens, shallowy 1`] = `
<div
  className="relative inline-block"
>
  <Button
    onClick={[Function]}
    theme="brand"
  >
    Test Button
  </Button>
  <ul
    className="w-full bg-gray-200 rounded absolute shadow-md invisible"
  >
    <DropdownItem>
      Item 1
    </DropdownItem>
    <DropdownItem>
      Item 2
    </DropdownItem>
    <DropdownItem>
      Item 3
    </DropdownItem>
  </ul>
</div>
`;

exports[`dropdown component render tests should render correctly without childrens, shallowy 1`] = `
<div
  className="relative inline-block"
>
  <Button
    onClick={[Function]}
    theme="brand"
  >
    Test Button
  </Button>
  <ul
    className="w-full bg-gray-200 rounded absolute shadow-md invisible"
  />
</div>
`;
